### PR TITLE
CNV-9002: Added module for creating a custom boot source

### PR DESCRIPTION
**Pull request by bgaydosrh**
For 4.8 only.

Jira: https://issues.redhat.com/browse/CNV-9002

Tagging: @yaacov for Code Review
Tagging: @gouyang for QE Review

Direct doc preview link: https://deploy-preview-33656--osdocs.netlify.app/openshift-enterprise/latest/virt/virtual_machines/virtual_disks/virt-creating-and-using-boot-sources.html#virt-creating-a-custom-disk-image-boot-source-web_virt-creating-and-using-boot-sources

**Questions and comments:** **(reviewers please respond)**

1. Some of the labels on the UI were obscured by the video software hotkey buttons used when the video was shot. Please verify the labels that I documented are correct. The one I know is NOT correct is **Boot source available** --- should this be "**Make boot source available**?" Similarly, on the Prepare boot source customization page, it appears the button label says **Start customization** but again I can't be sure.
2. Does a user need any specific admin privileges to perform this flow? It didn't appear so.It looks like temporary admin privileges are in effect during this process.
3. I tried to keep the names for the UI elements similar to those used in the RHEL Image Prep module that also appears in this assembly (second section). 
4. The reason I placed the Custom flow at the bottom is because, unlike the other modules in this assembly, the Custom flow  takes the user through (almost) the entire process, as opposed to the other modules which describe the three tasks in a standalone way (Sections 2-4: Image prep for RHEL, attaching the boot source to a template, creating the VM from the template). Should I add an instruction at the end of the custom flow to tell the user how to Create the VM from the custom template or does it seem they will find that module. It is immediately above, but it might not hurt to direct new users.
5. There appeared a slight lag in the video as the image starts up on the console, before you login. Let me know if there is something I can add beyond what I have documented.

Thanks, overall the video was very helpful.

Bob

### CODE REVIEW

File: modules/virt-creating-a-custom-image-boot-source-web.adoc [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+Sometimes you want to prepare a custom disk image, based on an existing disk image, for use as a boot source.
+
+This procedure guides you through:
+
+* Preparing a custom disk image.
+* Creating a boot source from the custom disk image.
+* Attaching the boot source to a custom template.
+
+.Prerequisites
+* You must have access to the web server with the operating system image you want to customize.
```
Comment (yaacov) [inline modules/virt-creating-a-custom-image-boot-source-web.adoc:17]:
```suggestion
```
You can customize locally stored images, access via web access is not required.

Comment (bgaydosrh) [reply modules/virt-creating-a-custom-image-boot-source-web.adoc:17]: Removed Prereq.

**Review by yaacov — APPROVED** [review]

**Comment by yaacov** [comment]
LGTM

File: modules/virt-creating-a-custom-image-boot-source-web.adoc [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+* Attaching the boot source to a custom template.
+
+.Procedure
+. In the {VirtProductName} console, click *Workloads* > *Virtualization* from the side menu.
+. Click the *Templates* tab.
+. Click the link in the *Source provider* column for the template you want to customize. A window displays, indicating that the template currently has a *defined* _source_.
+. In the window, click the *Customize source* link.
+. In the *About boot source customization* window, click *Continue* to proceed with customization, after reading the information provided about the boot source customization process.
+. On the *Prepare boot source customization* page, in the *Define new template* section:
+.. Select the *New template namespace* field and choose *PR default*.
```
Comment (gouyang) [inline modules/virt-creating-a-custom-image-boot-source-web.adoc:23]: It hasn't to be 'default' and `PR` is an abbreviation. Suggestion:
Click the *New template namespace* field and select a project.

**Comment by gouyang** [comment]
Others are LGTM.

File: modules/virt-creating-a-custom-image-boot-source-web.adoc (hunk 2 above)
Comment (bgaydosrh) [reply modules/virt-creating-a-custom-image-boot-source-web.adoc:23]: Thanks @gouyang. Before I make this change, can you also confirm that labels (in **Bold**) in these steps:

- _Verify the custom image works as expected. If it does, click **Boot source available**._
- 
- _Click **Start Customization**. The customization process begins and the Preparing boot source customization page displays, followed by the Customize boot source page. The Customize boot source page displays the output of the running script. When the script completes, your custom image is created._

"Boot source available" doesn't seem right to me, and I had a hard time seeing the label in the video. Just want to be sure, as I can't access a build.

File: modules/virt-creating-a-custom-image-boot-source-web.adoc [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+.. Select the *New template support* field and choose the appropriate value, indicating support contacts for the custom template you create.
+.. Select the *New template flavor* field and choose the appropriate CPU and memory values for the custom image you create.
+. In the *Prepare boot source for customization* section, customize the `cloud-init` YAML script, if needed, to define login credentials. Otherwise, default credentials will be automatically generated for you.
+. Click *Start Customization*. The customization process begins and the *Preparing boot source customization* page displays, followed by the *Customize boot source* page. The *Customize boot source* page displays the output of the running script. When the script completes, your custom image is created.
+. In the *VNC console*, click *show password* in the *Guest login credentials* section. Your login credentials are displayed.
+. When the image is ready for login, sign in with the *VNC Console* by providing the user name and password displayed in the *Guest login credentials* section.
+. Verify the custom image works as expected. If it does, click *Boot source available*.
```
Comment (gouyang) [inline modules/virt-creating-a-custom-image-boot-source-web.adoc:32]: it's `Make this boot source available` on this window.

File: modules/virt-creating-a-custom-image-boot-source-web.adoc (hunk 2 above)
Comment (gouyang) [reply modules/virt-creating-a-custom-image-boot-source-web.adoc:23]: You're right, it's `Make this boot source available`.

Comment (bgaydosrh) [reply modules/virt-creating-a-custom-image-boot-source-web.adoc:23]: Great! Thanks, fixed.

File: modules/virt-creating-a-custom-image-boot-source-web.adoc (hunk 3 above)
Comment (bgaydosrh) [reply modules/virt-creating-a-custom-image-boot-source-web.adoc:32]: Done.

File: modules/virt-creating-a-custom-image-boot-source-web.adoc (hunk 2 above)
Comment (bgaydosrh) [reply modules/virt-creating-a-custom-image-boot-source-web.adoc:23]: I'm not sure whether it should be click or select on the drop-downs, but peer review will set me straight on that. Changing to: Select the *New template namespace* field and choose a project.

File: modules/virt-creating-a-custom-image-boot-source-web.adoc [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+This procedure guides you through:
+
+* Preparing a custom disk image.
+* Creating a boot source from the custom disk image.
+* Attaching the boot source to a custom template.
+
+.Procedure
+. In the {VirtProductName} console, click *Workloads* > *Virtualization* from the side menu.
+. Click the *Templates* tab.
+. Click the link in the *Source provider* column for the template you want to customize. A window displays, indicating that the template currently has a *defined* _source_.
```
Comment (adellape) [inline modules/virt-creating-a-custom-image-boot-source-web.adoc:19]: Suggest dropping the emphasis on `_source_`, unless there's a meaning i'm not getting.

File: modules/virt-creating-a-custom-image-boot-source-web.adoc [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+. In the window, click the *Customize source* link.
+. In the *About boot source customization* window, click *Continue* to proceed with customization, after reading the information provided about the boot source customization process.
```
Comment (adellape) [inline modules/virt-creating-a-custom-image-boot-source-web.adoc:21]: Maybe reorder the sentence so the "reading" part is first:

```
. In the *About boot source customization* window, read the information provided about the boot source customization process, and then click *Continue* to proceed with customization.
```

File: modules/virt-creating-a-custom-image-boot-source-web.adoc [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+.. Enter the name of the template provider in the *New template provider* field.
+.. Select the *New template support* field and choose the appropriate value, indicating support contacts for the custom template you create.
+.. Select the *New template flavor* field and choose the appropriate CPU and memory values for the custom image you create.
+. In the *Prepare boot source for customization* section, customize the `cloud-init` YAML script, if needed, to define login credentials. Otherwise, the script generates default credentials for you.
+. Click *Start Customization*. The customization process begins and the *Preparing boot source customization* page displays, followed by the *Customize boot source* page. The *Customize boot source* page displays the output of the running script. When the script completes, your custom image is available.
+. In the *VNC console*, click *show password* in the *Guest login credentials* section. Your login credentials display.
+. When the image is ready for login, sign in with the *VNC Console* by providing the user name and password display in the *Guest login credentials* section.
```
Comment (adellape) [inline modules/virt-creating-a-custom-image-boot-source-web.adoc:31]: s/display/displayed/

?

File: modules/virt-creating-a-custom-image-boot-source-web.adoc (hunk 5 above)
Comment (adellape) [reply modules/virt-creating-a-custom-image-boot-source-web.adoc:21]: Also in general, I didn't point it out in the various other places that were similar, but something to consider is what IBM SG says about "then (_adverb_)" in the "Word usage" glossary, regarding two-step procedures written as one sentence. Technically, they should be joined by ", and then..." :shrug:

File: modules/virt-creating-a-custom-image-boot-source-web.adoc [hunk 7]
```diff
@@ -0,0 +1,35 @@
+// Module included in the following assemblies:
+//
+// * virt/virtual_machines/virtual_disks/virt-creating-and-using-boot-sources.adoc
+
+[id="virt-creating-a-custom-disk-image-boot-source-web_{context}"]
+= Creating a custom boot source
+
+Sometimes you want to prepare a custom disk image, based on an existing disk image, for use as a boot source.
```
[the remark this answers is not in the document]
Comment (bgaydosrh) [reply modules/virt-creating-a-custom-image-boot-source-web.adoc:8]: Done

File: modules/virt-creating-a-custom-image-boot-source-web.adoc [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+This procedure guides you through:
```
[the remark this answers is not in the document]
Comment (bgaydosrh) [reply modules/virt-creating-a-custom-image-boot-source-web.adoc:10]: OK

File: modules/virt-creating-a-custom-image-boot-source-web.adoc [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+* Preparing a custom disk image.
```
[the remark this answers is not in the document]
Comment (bgaydosrh) [reply modules/virt-creating-a-custom-image-boot-source-web.adoc:12]: Got it.

File: modules/virt-creating-a-custom-image-boot-source-web.adoc (hunk 4 above)
Comment (bgaydosrh) [reply modules/virt-creating-a-custom-image-boot-source-web.adoc:19]: The source is technically a variable, but I find it confusing too, so I'm dropping it. It's not really helping anything.

File: modules/virt-creating-a-custom-image-boot-source-web.adoc (hunk 5 above)
Comment (bgaydosrh) [reply modules/virt-creating-a-custom-image-boot-source-web.adoc:21]: Reworded and added the "thens" in other places where needed.

File: modules/virt-creating-a-custom-image-boot-source-web.adoc (hunk 6 above)
Comment (bgaydosrh) [reply modules/virt-creating-a-custom-image-boot-source-web.adoc:31]: Gak. Fixed.

**Review by adellape — APPROVED** [review]

**Pull request merged**